lib/llhttp/parser.js

```diff
@@ -0,0 +1,261 @@
+  i[kHeadersValues].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_headers_complete = p => {
+  const i = instMap.get(p)
+  const type = inst.exports.llhttp_get_type(p)
+  const versionMajor = inst.exports.llhttp_get_http_major(p)
+  const versionMinor = inst.exports.llhttp_get_http_minor(p)
+  const rawHeaders = i[kGetHeaders]()
+  let method
+  let url
+  let statusCode
+  let statusMessage
+  const upgrade = Boolean(inst.exports.llhttp_get_upgrade(p))
+  const shouldKeepAlive = inst.exports.llhttp_should_keep_alive(p)
+
+  // TODO: understand better when this is called in node parser
+  // and if it's needed. For now it causes duplicated header values
+  // i[kOnHeaders](rawHeaders)
+
+  if (type === HTTPParser.REQUEST) {
+    method = constants.METHODS[inst.exports.llhttp_get_method(p)]
+    url = i[kUrl]
+  } else if (type === HTTPParser.RESPONSE) {
+    statusCode = inst.exports.llhttp_get_status_code(p)
+    statusMessage = i[kStatusMessage]
+  }
+  return i[kOnHeadersComplete](versionMajor, versionMinor, rawHeaders, method,
+    url, statusCode, statusMessage, upgrade, shouldKeepAlive)
+}
+
+const wasm_on_body = (p, at, length) => {
+  const i = instMap.get(p)
+  const u8 = new Uint8Array(inst.exports.memory.buffer, at, length)
+  const body = Buffer.from(u8)
+  i[kOnBody](body)
+  return 0
+}
+
+const wasm_on_message_complete = (p) => {
+  const i = instMap.get(p)
+  // Handle trailers
+  if (i[kHeadersFields].length) {
+    const rawHeaders = i[kGetHeaders]()
+    i[kOnHeaders](rawHeaders)
+  }
+  return instMap.get(p)[kOnMessageComplete]()
+}
+
+/* eslint-enable camelcase */
+const instMap = new Map()
+
+const inst = new WebAssembly.Instance(mod, {
+  env: {
+    wasm_on_message_begin,
+    wasm_on_url,
+    wasm_on_status,
+    wasm_on_header_field,
+    wasm_on_header_value,
+    wasm_on_headers_complete,
+    wasm_on_body,
+    wasm_on_message_complete
+  }
+})
+
+inst.exports._initialize() // wasi reactor
+
+class HTTPParserError extends Error {
+  constructor (message, code) {
+    super(message)
+    Error.captureStackTrace(this, HTTPParserError)
+    this.name = 'HTTPParserError'
+    this.code = code
+  }
+}
+
+class HTTPParser {
+  constructor () {
+    this[kPtr] = null
+    this[kUrl] = ''
+    this[kStatusMessage] = null
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    this[kBody] = null
+  }
+
+  initialize (type) {
+    this[kPtr] = inst.exports.llhttp_alloc(constants.TYPE[type])
+    instMap.set(this[kPtr], this)
+  }
+
+  [kGetHeaders] () {
+    const rawHeaders = []
+    for (let c = 0; c < this[kHeadersFields].length; c++) {
+      rawHeaders.push(this[kHeadersFields][c], this[kHeadersValues][c])
+    }
+    this[kHeadersFields] = []
+    this[kHeadersValues] = []
+    return rawHeaders
+  }
+
+  [kReset] () {
+    this[kUrl] = ''
+    this[kStatusMessage] = null
+    this[kBody] = null
+  }
+
+  [kOnMessageBegin] () {
+    return 0
+  }
+
+  [kOnHeaders] (rawHeaders) {}
```

Comment:
not needed

Reply:
This one is needed to handle trailers, or you mean there's no need for an empty/virtual method?

Ref: https://github.com/nodejs/undici/blob/parser-execute/lib/core/client.js#L437